pkg/controller/velero/controller.go

```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to BackupStorageLocation
+	err = c.Watch(&source.Kind{Type: &velerov1.BackupStorageLocation{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to VolumeSnapshotLocation
+	err = c.Watch(&source.Kind{Type: &velerov1.VolumeSnapshotLocation{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to CredentialsRequest
+	err = c.Watch(&source.Kind{Type: &minterv1.CredentialsRequest{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to Deployments
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &veleroCR.Velero{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileVelero implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileVelero{}
+
+// ReconcileVelero reconciles a Velero object
+type ReconcileVelero struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Velero object and makes changes based on the state read
+// and what is in the Velero.Spec
+func (r *ReconcileVelero) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling Velero Installation")
+	var err error
+
+	// Fetch the Velero instance
+	instance := &veleroCR.Velero{}
+	err = r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Grab platform status to determine where OpenShift is installed
+	platformStatus, err := platform.GetPlatformStatus(r.client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Verify that we have received an AWS region from the platform
+	if platformStatus.AWS == nil || len(platformStatus.AWS.Region) < 1 {
+		return reconcile.Result{}, fmt.Errorf("unable to determine AWS region")
+	}
+
+	// Create an S3 client based on the region we received
+	s3Client, err := s3.NewS3Client(r.client, platformStatus.AWS.Region)
+	if err != nil {
+		return reconcile.Result{}, err
```

Comment:
Do you think logging potential aws errors here would be helpful?

Reply:
Yes, you're right.. I'm not printing many of these out to the terminal. Will add that!

Reply:
Oh, right.. so I had tested this. When the `Reconcile` function passes up to the controller a non-nil err, the controller itself will log out the err:
```
{"level":"error","ts":1566589629.536369,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"velero-controller","request":"openshift-velero/cluster","error":"unable to determine AWS region","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\tpkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:217\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.1.12/pkg/internal/controller/controller.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\tpkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\tpkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:134\nk8s.io/apimachinery/pkg/util/wait.Until\n\tpkg/mod/k8s.io/apimachinery@v0.0.0-20190221213512-86fb29eff628/pkg/util/wait/wait.go:88"}
```